integration/test/Test/MLS/KeyPackage.hs

```diff
@@ -68,10 +66,79 @@ testKeyPackageCount cs = do
 
 testUnsupportedCiphersuite :: HasCallStack => App ()
 testUnsupportedCiphersuite = do
-  setMLSCiphersuite (Ciphersuite "0x0002")
+  let suite = Ciphersuite "0x0002"
+  setMLSCiphersuite suite
   bob <- randomUser OwnDomain def
   bob1 <- createMLSClient def bob
-  (kp, _) <- generateKeyPackage bob1
+  (kp, _) <- generateKeyPackage bob1 suite
   bindResponse (uploadKeyPackages bob1 [kp]) $ \resp -> do
     resp.status `shouldMatchInt` 400
     resp.json %. "label" `shouldMatch` "mls-protocol-error"
+
+testReplaceKeyPackages :: HasCallStack => App ()
+testReplaceKeyPackages = do
+  alice <- randomUser OwnDomain def
+  [alice1, alice2] <- replicateM 2 $ createMLSClient def alice
+  let suite = Ciphersuite "0xf031"
+
+  let checkCount cs n =
+        bindResponse (countKeyPackages cs alice1) $ \resp -> do
+          resp.status `shouldMatchInt` 200
+          resp.json %. "count" `shouldMatchInt` n
+
+  -- setup: upload a batch of key packages for each ciphersuite
+  void $
+    replicateM 4 (fmap fst (generateKeyPackage alice1 def))
+      >>= uploadKeyPackages alice1
+      >>= getBody 201
+  setMLSCiphersuite suite
+  void $
+    replicateM 5 (fmap fst (generateKeyPackage alice1 suite))
+      >>= uploadKeyPackages alice1
+      >>= getBody 201
+
+  checkCount def 4
+  checkCount suite 5
+
+  do
+    -- generate a new batch of key packages
+    (kps, refs) <- unzip <$> replicateM 3 (generateKeyPackage alice1 suite)
+
+    -- replace old key packages with new
+    void $ replaceKeyPackages alice1 [suite] kps >>= getBody 201
+
+    checkCount def 4
+    checkCount suite 3
+
+    -- claim all key packages one by one
+    claimed <-
+      replicateM 3 $
+        bindResponse (claimKeyPackages suite alice2 alice) $ \resp -> do
+          resp.status `shouldMatchInt` 200
+          ks <- resp.json %. "key_packages" & asList
+          k <- assertOne ks
+          k %. "key_package_ref"
+
+    refs `shouldMatchSet` claimed
+
+    checkCount def 4
+    checkCount suite 0
+
+  when False $ do
```

Comment:
This never runs, no?

Reply:
Oops :sweat_smile:. Thanks for catching that.